### ISSUE TITLE
Bug: not deterministic reporting of `invalid-return-type`

### ISSUE THREAD

**Issue by jeertmans** [astral-sh/ty#2507]
### Summary

Hello!

I am not sure how to report this, but it appears that the output of `ty check` is not deterministic. I discovered this when trying to ignore a specific `invalid-return-type` diagnostic, that I didn't know how to fix. I enabled the `unused-ignore-comment` rule, and noticed that the rule would sometimes be triggered, sometimes not.

```
➜  DiffeRT git:(main) ✗ uv run ty check
All checks passed!
➜  DiffeRT git:(main) ✗ uv run ty check
error[unused-ignore-comment]: Unused `ty: ignore` directive
   --> differt/src/differt/plotting/_utils.py:401:33
    |
399 |             registry[backend] = __wrapper__
400 |
401 |             return __wrapper__  # ty: ignore[invalid-return-type]
    |                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
402 |
403 |         return _wrapper_
    |
help: Remove the unused suppression comment
398 | 
399 |             registry[backend] = __wrapper__
400 | 
    -             return __wrapper__  # ty: ignore[invalid-return-type]
401 +             return __wrapper__
402 | 
403 |         return _wrapper_
404 | 

Found 1 diagnostic
➜  DiffeRT git:(main) ✗ uv run ty check
error[unused-ignore-comment]: Unused `ty: ignore` directive
   --> differt/src/differt/plotting/_utils.py:401:33
    |
399 |             registry[backend] = __wrapper__
400 |
401 |             return __wrapper__  # ty: ignore[invalid-return-type]
    |                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
402 |
403 |         return _wrapper_
    |
help: Remove the unused suppression comment
398 | 
399 |             registry[backend] = __wrapper__
400 | 
    -             return __wrapper__  # ty: ignore[invalid-return-type]
401 +             return __wrapper__
402 | 
403 |         return _wrapper_
404 | 

Found 1 diagnostic
➜  DiffeRT git:(main) ✗ uv run ty check
error[unused-ignore-comment]: Unused `ty: ignore` directive
   --> differt/src/differt/plotting/_utils.py:401:33
    |
399 |             registry[backend] = __wrapper__
400 |
401 |             return __wrapper__  # ty: ignore[invalid-return-type]
    |                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
402 |
403 |         return _wrapper_
    |
help: Remove the unused suppression comment
398 | 
399 |             registry[backend] = __wrapper__
400 | 
    -             return __wrapper__  # ty: ignore[invalid-return-type]
401 +             return __wrapper__
402 | 
403 |         return _wrapper_
404 | 

Found 1 diagnostic
➜  DiffeRT git:(main) ✗ uv run ty check
All checks passed!
```

Removing the `# ty: ignore[invalid-return-type]` comment does the opposite:

```
➜  DiffeRT git:(main) ✗ uv run ty check
error[invalid-return-type]: Return type does not match returned value
   --> differt/src/differt/plotting/_utils.py:401:20
    |
399 |             registry[backend] = __wrapper__
400 |
401 |             return __wrapper__
    |                    ^^^^^^^^^^^ expected `(**P@dispatch) -> T@dispatch`, found `_Wrapped[P@dispatch, T@dispatch | SceneCanvas | matplotlib.figure.Figure | plotly.graph_objs._figure.Figure, P@dispatch, T@dispatch | SceneCanvas | matplotlib.figure.Figure | plotly.graph_objs._figure.Figure]`
402 |
403 |         return _wrapper_
    |
   ::: differt/src/differt/plotting/_utils.py:381:48
    |
379 |             )
380 |
381 |         def _wrapper_(impl: Callable[P, T]) -> Callable[P, T]:
    |                                                -------------- Expected `(**P@dispatch) -> T@dispatch` because of return type
382 |             # ruff: noqa: DOC201
383 |             """Actually register the backend implementation."""
    |
info: rule `invalid-return-type` is enabled by default

Found 1 diagnostic
➜  DiffeRT git:(main) ✗ uv run ty check
All checks passed!
➜  DiffeRT git:(main) ✗ uv run ty check
error[invalid-return-type]: Return type does not match returned value
   --> differt/src/differt/plotting/_utils.py:401:20
    |
399 |             registry[backend] = __wrapper__
400 |
401 |             return __wrapper__
    |                    ^^^^^^^^^^^ expected `(**P@dispatch) -> T@dispatch`, found `_Wrapped[P@dispatch, T@dispatch | SceneCanvas | matplotlib.figure.Figure | plotly.graph_objs._figure.Figure, P@dispatch, T@dispatch | SceneCanvas | matplotlib.figure.Figure | plotly.graph_objs._figure.Figure]`
402 |
403 |         return _wrapper_
    |
   ::: differt/src/differt/plotting/_utils.py:381:48
    |
379 |             )
380 |
381 |         def _wrapper_(impl: Callable[P, T]) -> Callable[P, T]:
    |                                                -------------- Expected `(**P@dispatch) -> T@dispatch` because of return type
382 |             # ruff: noqa: DOC201
383 |             """Actually register the backend implementation."""
    |
info: rule `invalid-return-type` is enabled by default

Found 1 diagnostic
```

Do you have any idea if this is a bug from ty?

# MWE

You can reproduce this bug by cloning my repository:

```bash
git clone https://github.com/jeertmans/DiffeRT.git --depth 1
git reset --hard e031efca11c63a5f7e9e75ee02a025451beb2f8e
```

You can avoid building the Rust code by downloading pre-built wheels:

```bash
sed -i 's/members = \["differt", "differt-core"\]/members = ["differt"]/g' pyproject.toml
sed -i 's/^differt_core/# differt_core/g' differt/pyproject.toml
```

### Version

ty 0.0.12

**Comment by carljm** [comment]
Thanks for the report! Yes, we have some known non-determinism on some projects that we need to track down and fix; another example is helpful.

**Comment by jeertmans** [comment]
Glad to hear I am not the only one :D Let me know if I can help

**Comment by carljm** [comment]
Making some notes from investigation so far. It looks like the cause of this issue is creation of union types with elements inserted in different order depending on query execution order. This union type creation may happen in `cycle_normalized`, but not necessarily -- exploration of the DiffeRT repro provided here suggests that the problematic union is actually created in `SpecializationBuilder::add_type_mappings_from_constraint_set` in this case.

In principle our `UnionBuilder` simplification of unions should not be sensitive to the order elements are inserted in the union, but this non-sensitivity to order depends on transitivity of subtyping, and we know that this transitivity does not always hold in practice. In particular we know that transitivity of subtyping with respect to structural types is broken by nominal inheritance with Liskov violation. For example:

```py
from typing import Protocol

class A:
    x: int

class B(A):
    x: str  # Liskov violation

class P(Protocol):
    x: int
```

With these types, we have that `B <: A` (by nominal inheritance), `A <: P` (by structural subtyping), but `B </: P` (by structural subtyping), breaking transitivity.

(I haven't confirmed that this particular form of non-transitivity is the cause of the issue in DiffeRT specifically -- this just confirms that we have known cases of non-transitivity.)

In order to solve the non-determinism, we likely need to make `UnionBuilder` results not sensitive to the order in which elements are inserted. If we are not able to ensure transitivity of subtyping, this likely requires a different approach to union simplification (e.g. doing it lazily at `finish()` rather than eagerly as each element is inserted, and iterating the elements in some stable order when doing simplification.)

(cc @mtshiba )

**Comment by carljm** [comment]
> a different approach to union simplification (e.g. doing it lazily at `finish()` rather than eagerly as each element is inserted, and iterating the elements in some stable order when doing simplification.)

It may be too expensive to do this in general, as it could require allocating large amounts of memory for collecting elements that today would be eagerly simplified out early.

If we are able to identify the specific cases of non-transitivity that cause problems, and it is structural types that are the problem, we could address that in a more targeted way, e.g. by doing this lazy simplification only for structural types? Or even go further and don't ever simplify types out of unions due to redundancy with structural types? (This could also help address our issues with `Hashable` and `object`.)

**Comment by carljm** [comment]
The other direction we could go here is to try to restore and enforce transitivity of subtyping, but the options here all seem likely to cause even worse problems. In principle a Liskov-violating attribute of a subclass could be inferred as the intersection with the base-class attribute type, or as `Unknown`, preserving Liskov -- but Liskov violations are unfortunately not uncommon, and this inference would definitely surprise users.

**Comment by MichaReiser** [comment]
> It may be too expensive to do this in general, as it could require allocating large amounts of memory for collecting elements that today would be eagerly simplified out early.

What TypeScript does is it performs a "cheap" (and I believe linear) simplification on insertion, which roughly resembles what we do for literals today. The full simplification is done lazily and cached. 

I did implement this once and the performance was worse (but I doubt it was like 2x, more like 10% or something) but I fail to find my PR. The way I implemented it is that `UnionType` had a salsa cached method that returns the lazily simplified union.

TypeScript's union simplification: https://github.com/microsoft/TypeScript/blob/60934915d9ccc4ca9c0fb2cd060d7ec81601942b/src/compiler/checker.ts#L14948

Subtyping part: https://github.com/microsoft/TypeScript/blob/60934915d9ccc4ca9c0fb2cd060d7ec81601942b/src/compiler/checker.ts#L14818-L14882

**Comment by carljm** [comment]
Given the above non-transitive types `A`, `B`, and `P`, this snippet demonstrates concretely how the results of our current union simplification depend on the order of element insertion. If we create a union of all three types, depending on the order we insert them, the result may be just `P`, or may be `P | B` (or `B | P`):

```py
def _(
    ab: A | B,
    ap: A | P,
    bp: B | P,
    abp: A | B | P,
    apb: A | P | B,
    bap: B | A | P,
    bpa: B | P | A,
    pab: P | A | B,
    pba: P | B | A,
):
    reveal_type(ab)  # A
    reveal_type(ap)  # P
    reveal_type(bp)  # B | P
    reveal_type(abp) # P
    reveal_type(apb) # P | B
    reveal_type(bap) # P
    reveal_type(bpa) # B | P
    reveal_type(pab) # P | B
    reveal_type(pba) # P | B
```

**Comment by carljm** [comment]
> `UnionType` had a salsa cached method that returns the lazily simplified union.

This is more lazy than what I proposed above. What I proposed would be lazy only within `UnionBuilder` itself, but would still simplify on `UnionBuilder::build()`, it would not intern the full non-simplified collection of types. I think the latter would be sufficient to solve this problem. It also shouldn't be that complex or invasive to implement, since the actual simplification logic shouldn't need to change much. It would require we can identify a "stable" order in which to iterate the types for simplification. And it could still mean that we need a larger (maybe sometimes much larger) temporary allocation within `UnionBuilder`.

**Comment by AlexWaygood** [comment]
> In principle our `UnionBuilder` simplification of unions should not be sensitive to the order elements are inserted in the union, but this non-sensitivity to order depends on transitivity of subtyping, and we know that this transitivity does not always hold in practice. In particular we know that transitivity of subtyping with respect to structural types is broken by nominal inheritance with Liskov violation.

This working theory for the cause of the nondeterminism is appealing, because it goes a long way towards explaining why the number of nondeterministic diagnostics would have spiked after https://github.com/astral-sh/ruff/pull/21551 was merged: `Callable` types are structural types

**Comment by carljm** [comment]
It's also the case that a lot of the diagnostics that disappear and appear seem to involve smaller vs larger union types.

**Comment by MichaReiser** [comment]
> (e.g. doing it lazily at finish() rather than eagerly as each element is inserted, and iterating the elements in some stable order when doing simplification.)

I obviously don't have deep knowledge on this, so I might be completely off but:

What would you use as a stable ordering? I don't think it can be the source position, because that would mean that reordering elements in a union could introduce new errors in entirely different parts of your code bases. But that means it can't be the type's qualified name, or renaming a type could introduce new type errors if it is used in a union, nor can it be its structure, because adding or removing a method could then lead to new errors. Or is the idea that we sort by type kind? 

Maybe put differently. If we have a liskov violation, what's the expected type of getting the attribute `x` from a union over `P`, `B` and `A`? Is it `str | int` or `int`?

**Comment by carljm** [comment]
> I don't think it can be the source position, because that would mean that reordering elements in a union could introduce new errors in entirely different parts of your code bases.

I think that would still be strictly better behavior than being non-deterministic! But I agree it's not ideal.

> Maybe put differently. If we have a liskov violation, what's the expected type of getting the attribute `x` from a union over `P`, `B` and `A`? Is it `str | int` or `int`?

Pyright and pyrefly both seem to simplify unions rarely if ever, and say `int | str`. Mypy simplifies the union if it comes from control flow (e.g. `x = A() if flag else B()`) and in that case reveals just `int`, but does not simplify the union (and reveals `int | str`) if there's an explicit `A | B | P` or `A | B` annotation.

I don't think there's a clear right answer here. The only answer from theory is "you shouldn't have Liskov violations", and that would imply that we actually convert the `x: str` on `B` to either `Unknown` or `int & str` (which would become `Never` in this case). But that would definitely surprise people, and would violate the "respect explicit annotations" principle. Otherwise, it just becomes a question of what's more useful in practice. I think probably that's `int | str` (at least in the case where the "redundant" union is explicitly written out in an annotation), which would imply that the direction we should go to resolve this is "simplify unions less." But I'm not sure how much that would help, because I think we need to generally simplify unions in places like `cycle_normalized` and `SpecializationBuilder` where the unions are fully synthetic. (Though maybe even there, there are only some kinds of simplifications we need: `int` replacing int literals and similar.)

One option to consider is that we track a flag on any type that violates Liskov, and we don't simplify that type out of unions with its base(s).

**Comment by MichaReiser** [comment]
> I don't think there's a clear right answer here. The only answer from theory is "you shouldn't have Liskov violations", and that would imply that we actually convert the x: str on B to either Unknown or int & str (which would become Never in this case).

So, Liskov violations should become non-suppressable errors? :P 

> which would imply that the direction we should go to resolve this is "simplify unions less."

That's sort of the impression I'm getting, given that we can't enforce Liskov. Unless we say the simplification is okay on the same terms as is narrowing on `str` where we have a rule that warns you if you subclass `str` in a way that violates our assumptions. But given that Liskov violations are so common (they're in stdlib), I'm not sure it's a viable path, especially because it isn't necessarily a "simple" solution. 

I don't understand much of TypeScript's code but they seem to use different reductions based on context too, where subtyping is one possible reduction (https://github.com/microsoft/TypeScript/blob/60934915d9ccc4ca9c0fb2cd060d7ec81601942b/src/compiler/checker.ts#L14818-L14882). There must be some reason why all other type checkers don't reduce eary and maybe we know learned why :)

**Comment by jeertmans** [comment]
Jumping into your discussion (very interesting!) to say that (as a user) I would very much prefer to have an error raised when there is a Liskov violation, and enforce myself to fix it or tell  ty to ignore it, than having ty trying to resolve the type of something I have clearly broken.

**Comment by carljm** [comment]
> Jumping into your discussion (very interesting!) to say that (as a user) I would very much prefer to have an error raised when there is a Liskov violation, and enforce myself to fix it or tell ty to ignore it, than having ty trying to resolve the type of something I have clearly broken.

We definitely intend to error on Liskov violations (and we already do for methods, but not yet for attributes); that's not in question. The discussion here is related to what we do with the Liskov-violating subclasses, when the Liskov violation diagnostic is suppressed (or just ignored). That's what we mean by "can't enforce Liskov". Even if we emit a diagnostic, that doesn't mean we can just stop there. We still have to type check the rest of the project, and we have to use _some_ type for the Liskov-violating override.
